Mathlib/Algebra/Module/LocalizedModule/Basic.lean

```diff
@@ -117,187 +129,122 @@ theorem liftOn_mk {α : Type*} {f : M × S → α} (wd : ∀ (p p' : M × S), p
+/--
+The multiplication on the localized module.
+Note that this gives a diamond with the instance on `R[S⁻¹]` (which does not require commutativity),
+but is defeq to it under `with_reducible_and_instances`.
+See https://github.com/leanprover-community/mathlib4/pull/25671 for an approach to generalize this
+but it requires right `R` actions on `R`-algebras.
+-/
+protected def mul {A : Type*} [Semiring A] [Algebra R A] {S : Submonoid R}
+    (m₁ m₂ : LocalizedModule S A) : LocalizedModule S A :=
+  liftOn₂ m₁ m₂ (fun x₁ x₂ => LocalizedModule.mk (x₁.1 * x₂.1) (x₂.2 * x₁.2)) (by
+    rintro ⟨a₁, s₁⟩ ⟨a₂, s₂⟩ ⟨b₁, t₁⟩ ⟨b₂, t₂⟩ ⟨u₁, e₁⟩ ⟨u₂, e₂⟩
+    simp only [mul_comm s₂ s₁, mul_comm t₂ t₁]
+    rw [mk_eq]
+    use u₁ * u₂
+    dsimp [Submonoid.smul_def] at *
+    simp only [mul_smul_mul_comm, e₁, e₂])
+
+instance (priority := 900) {A : Type*} [Semiring A] [Algebra R A] {S : Submonoid R} :
     Monoid (LocalizedModule S A) :=
-  { mul := fun m₁ m₂ =>
-      liftOn₂ m₁ m₂ (fun x₁ x₂ => LocalizedModule.mk (x₁.1 * x₂.1) (x₁.2 * x₂.2))
-        (by
-          rintro ⟨a₁, s₁⟩ ⟨a₂, s₂⟩ ⟨b₁, t₁⟩ ⟨b₂, t₂⟩ ⟨u₁, e₁⟩ ⟨u₂, e₂⟩
-          rw [mk_eq]
-          use u₁ * u₂
-          dsimp only at e₁ e₂ ⊢
-          rw [eq_comm]
-          trans (u₁ • t₁ • a₁) • u₂ • t₂ • a₂
-          on_goal 1 => rw [e₁, e₂]
-          on_goal 2 => rw [eq_comm]
-          all_goals
-            rw [smul_smul, mul_mul_mul_comm, ← smul_eq_mul, ← smul_eq_mul (α := A),
-              smul_smul_smul_comm, mul_smul, mul_smul])
-    one := mk 1 (1 : S)
+  fast_instance%
+  { __ := inferInstanceAs (One (LocalizedModule S A))
+    mul := LocalizedModule.mul
     one_mul := by
       rintro ⟨a, s⟩
-      exact mk_eq.mpr ⟨1, by simp only [one_mul, one_smul]⟩
+      with_unfolding_all exact mk_eq.mpr ⟨1, by simp only [one_mul, mul_one, one_smul]⟩
     mul_one := by
       rintro ⟨a, s⟩
-      exact mk_eq.mpr ⟨1, by simp only [mul_one, one_smul]⟩
-    mul_assoc := by
+      with_unfolding_all exact mk_eq.mpr ⟨1, by simp only [mul_one, one_smul, one_mul]⟩
+    mul_assoc := by with_unfolding_all
       rintro ⟨a₁, s₁⟩ ⟨a₂, s₂⟩ ⟨a₃, s₃⟩
       apply mk_eq.mpr _
       use 1
-      simp only [one_mul, smul_smul, ← mul_assoc] }
+      simp only [one_mul, smul_smul, ← mul_assoc, mul_right_comm] }
+
+example : OreLocalization.instMonoid = LocalizedModule.instMonoid (A := R) (S := S) := by
+  with_reducible_and_instances rfl
+
+/-- A variant of `mk_mul_mk` that is `rfl` but has a stranger multiplication order. -/
+theorem mk_mul_mk' {A : Type*} [Semiring A] [Algebra R A] {a₁ a₂ : A} {s₁ s₂ : S} :
+    mk a₁ s₁ * mk a₂ s₂ = mk (a₁ * a₂) (s₂ * s₁) := rfl
+
+theorem mk_mul_mk {A : Type*} [Semiring A] [Algebra R A] {a₁ a₂ : A} {s₁ s₂ : S} :
+    mk a₁ s₁ * mk a₂ s₂ = mk (a₁ * a₂) (s₁ * s₂) := by rw [mk_mul_mk', mul_comm s₁ s₂]
 
-instance {A : Type*} [Semiring A] [Algebra R A] {S : Submonoid R} :
+instance (priority := 900) {A : Type*} [Semiring A] [Algebra R A] {S : Submonoid R} :
```

Comment:
What are we avoiding?

Reply:
We prefer the `OreLocalization.instSemiring` instance on `Localization S`. I have added a comment explaining this.